||||
| --- | --- | --- |
| [← 2.2. Details of GIVE Web Components](2.2-webComponents.md) | [↑ Index](Readme.md) | [3.2. MySQL commands for data management →](3.2-dataSource.md) |

# GIVE-Toolbox usages for managing data in GIVE data source

* [1. initial_ref.sh:](#1-initial_refsh)
* [2. add_trackGroup.sh](#2-add_trackgroupsh)
* [3. add_geneAnnot.sh](#3-add_geneannotsh)
* [4. add_track_bed.sh](#4-add_track_bedsh)
* [5. add_track_bigWig.sh](#5-add_track_bigwigsh)
* [6. add_track_interaction.sh](#6-add_track_interactionsh)
* [7. config_host.sh](#7-config_hostsh)
* [8. list_tracks.sh](#8-list_trackssh)
* [9. remove_data.sh](#9-remove_datash)



### 1. `initial_ref.sh`:
Initialize GIVE database structure with a new reference genome.

```
bash initial_ref.sh [-u <mysqlu>] [-p <mysqlp>] [-r <ref>] [-s <species_name>] [-c <species_cname>] [-f <file>]
```
The arguments of `initial_ref.sh` are:
- `-u <mysqlu>`：(Required) MySQL account name. All the databases will be created under this account, such as `root` in GIVE-Docker. Make sure you have access to this account.
- `-p <mysqlp>`: (Optional) MySQL account password. Without '-p', the prompt will ask you to input password
- `-r <ref>`: (Required) ref genome database name. The database name of reference genome. It must be a valid name less than 10 characters in MySQL, i.e., it cannot contain space, `/`, `\`, `.`, and characters that are not allowed in file name. Example: `hg19`, `hg38`, `b37_decoy`.
- `-s <species_name>`: (Required) species name of the ref genome. The name of the species, such as scientific name, `Homo sapiens`. If it contains space or other special character, please quote it, such as "Homo sapiens".
- `-c <species_cname>`: (Required) common name of the species. Such as `human`. If it contains space or other special character, please quote it.
- `-f <file>`: (Required) cytoBandIdeo file path. It must be a system absolute path. The cytoBandIdeo file defines the genome coordinates. You can download it from UCSC Genome Browser. We also supply some cytoBandIdeo of common human/mouse genomes [here](https://demo.give.genemo.org/annotations/).

### 2. `add_trackGroup.sh`
Add a track group to a reference genome.

```
bash add_trackGroup.sh [-u <mysqlu>] [-p <mysqlp>] [-r <ref>] [-g <group_name>] [-l <long_label>] [-o <priority>] [-s <single_choice>]
```
- `-u <mysqlu>`: (Required) MySQL account name
- `-p <mysqlp>`: (Optional) MySQL account password. Without '-p', the prompt will ask you to input password.
- `-r <ref>`: (Required) ref genome database name. The ref genome database must has been initialized. If not, please use initial_ref.sh to do it.
- `-g <group_name>`: (Required) user defined track group name.
- `-l <long_label>`: (Required) The long label of group, which will be shown on the front end of GIVE.
- `-o <priority>`: (Required) Integer start from 0. It will determine the order of groups, 0 is the top.
- `-s <single_choice>`: (Required) 0 or 1. Whether the group will only allow one track to be active at any time. 0 is FALSE, 1 is TRUE.

### 3. `add_geneAnnot.sh`
Add UCSC known gene annotations.

```bash
bash add_geneAnnot.sh [-u <mysqlu>] [-p <mysqlp>] [-r <ref>][-t <track_name>] [-g <group_name>] [-l <long_label>] [-s<short_label>]
[-o <priority>] [-v <visibility>] [-f <file>]
```

- `-u <mysqlu>`: (required) MySQL account name
- `-p <mysqlp>`: (Optional) MySQL account password. Without '-p', the prompt - will ask you to input password.
- `-r <ref>`: (Required) ref genome database name. The ref genome database must has been initialized. If not, please use initial_ref.sh to do it.
- `-t <track_name>`: (Required) track name you want to build from your data.
- `-g <group_name>`: (Required) user defined track group name. Multiple tracks can be grouped based on experimental design or techniques, such as group "geneAnnotation", "HiC"and "RNAseq".
- `-l <long_label>`: (Required) The long label of track.
- `-s <short_label>`: (Required) The short label of track.
- `-o <priority>`: (Required) The order of the track in the browser. Smaller value means the the track will be shown in a higher location.
- `-v <visibility>`: (Required) "full", "pack", "collapsed", "notext", "dense", "none". Usually use "pack".
- `-f <file>`: (Required) gene annotation file path. It must be a system absolute path.
- `-h`: show usage help

### 4. `add_track_bed.sh`
Add a data track from `bed` format file to a track group.

```bash
bash add_track_bed.sh [-u <mysqlu>] [-p <mysqlp>] [-r <ref>][-t <track_name>] [-g <group_name>] [-l <long_label>] [-s<short_label>] [-o <priority>] [-v <visibility>] [-f <file>]
```

- `-u <mysqlu>`: (required) MySQL account name
- `-p <mysqlp>`: (Optional) MySQL account password. Without '-p', the prompt will ask you to input password.
- `-r <ref>`: (Required) ref genome database name. The ref genome database must has been initialized. If not, please use `initial_ref.sh` to do it.
- `-t <track_name>`: (Required) track name you want to build from your data.
- `-g <group_name>`: (Required) user defined track group name. Multiple tracks can be grouped based on experimental design or techniques, such  as group "HiC"and "RNAseq".
- `-l <long_label>`: (Required) The long label of track.
- `-s <short_label>`: (Required) The short label of track.
- `-o <priority>`: (Required) The order of the track in the browser. Smaller value means the the track will be shown in a higher location.
- `-v <visibility>`: (Required) "full", "pack", "collapsed", "notext", "dense", "none". Usually use "pack".
- `-f <file>`: (Required) bed format file. It must be a system absolute path.
- `-h`: show usage help



### 5. `add_track_bigWig.sh`
Use script `add_track_bigwig.sh` to build track from bigWig data file.
```
bash add_track_bigwig.sh [-u <mysqlu>] [-p <mysqlp>] [-r <ref>] [-t <track_name>] [-g <group_name>] [-l <long_label>] [-s <short_label>] [-o <priority>] [-v <visibility>] [-a <autoscale>] [-W <windowMax>] [-w <windowMin>] [-f <file>]
```

- `-u <mysqlu>`: (Required) MySQL account name
- `-p <mysqlp>`: (Optional) MySQL account password. Without '-p', the prompt will ask you to input password.
- `-r <ref>`: (Required) ref genome database name. The ref genome database must has been initialized. If not, please use initial_ref.sh to do it.
- `-t <track_name>`: (Required) track name you want to build from your data.
- `-g <group_name>`: (Required) user defined track group name. Multiple tracks can be grouped based on experimental design or techniques, such as group "HiC"and "RNAseq".
- `-l <long_label>`: (Required) More detailed description of the track. Will be shown in a future update.  
- `-s <short_label>`: (Required) The label that will be shown in the label region.
- `-o <priority>`: (Required) The order of the track in the browser. Smaller value means the the track will be shown in a higher location.
- `-v <visibility>`: (Required) "full" or "none". The display mode of the track. If "full", signals will be plotted against the genome in a line graph. If "none", this track is not shown at all.
- `-a <autoscale>`: (Required) "true" or "false". Whether the display window is scaled automatically. When this is set to true, the maximum and minimum value will be calculated to be the 95th percentile and 5th percentile of all visible data. If zero is not included in the range, the range will automatically be expanded to include zero.
- `-W <windowMax>`: (Optional) The maximum value of data shown in the window, only effective when autoScale is set to false.
- `-w <windowMin>`: (Optional) The minimum value of data shown in the window, only effective when autoScale is set to false.
- `-f <file>`: (Required) bigWig file path. It must be a system absolute path. Make sure MySQL can access this file.


### 6. `add_track_interaction.sh`

```
bash add_track_interaction.sh [-u <mysqlu>] [-p <mysqlp>] [-r <ref>] [-t <track_name>] [-g <group_name>] [-l <long_label>] [-s <short_label>] [-o <priority>] [-v <visibility>] [-q <quantile>] [-f <file>]
```
- `-u <mysqlu>`: (Required) MySQL account name
- `-p <mysqlp>`: (Optional) MySQL account password. Without '-p', the prompt will ask you to input password.
- `-r <ref>`: (Required) ref genome database name. The ref genome database must has been initialized. If not, please use initial_ref.sh to do it.
- `-t <track_name>`: (Required) track name you want to build from your data.
- `-g <group_name>`: (Required) user defined track group name. Multiple tracks can be grouped based on experimental design or techniques, such as group "HiC"and "RNAseq".
- `-l <long_label>`: (Required) More detailed description of the track. Will be shown in a future update.  
- `-s <short_label>`: (Required) The label that will be shown in the label region.
- `-o <priority>`: (Required) The order of the track in the browser. Smaller value means the the track will be shown in a higher location.
- `-v <visibility>`: (Required) "full" or "none". The display mode of the track. If "full", signals will be plotted against the genome in a line graph. If "none", this track is not shown at all.
- `-q <quantiles>`: (Optional) The quantile values used for scaling. If color gradient based on quantile instead of absolute values is desired, an array of quantile values should be provided here. value will be scaled by quantiles before being mapped onto the color gradient. An example: "0.37,1.32,1.78,2.19,2.60,2.97,3.43,3.85,4.34,4.90,5.48,6.16,6.94,8.01,9.05,10.41,12.37,14.88,19.84,31.77,290.17"
- `-f <file>`: (Required) interaction file path. It must be a system absolute path. Make sure MySQL can access this file.

### 7. `config_host.sh`
This script tool will configure the "<give_root>/includes/constants.php" and "<give_root>/html/components/basic-func/constants.js" with user supplied arguments. Please make sure that you have the authority to write these files.

```bash
bash config_host.sh [-r <give_root>] [-u <mysqlu>] [-p <mysqlp>] [-d <host_domain>]
```

- `-r <give_root>`: (Required) The root directory of GIVE. The default dir in our tutorial is "/var/www/give".
- `-u <mysqlu>`: (Optional) MySQL account name. It must be correctly set or you will not get any track data. If supplied, it will change the file "<give_root>/includes/constants.php"
- `-p <mysqlp>`: (Optional) MySQL account password. It must be correctly set or you will not get any track data. If supplied, it will change the file "<give_root>/html/components/basic-func/constants.js"
- `-d <host_domain>`: (Optional) If the local machine has web service and you want to access GIVE through web service, you need to set <host_domain> in the file "<give_root>/html/components/basic-func/constants.js". <host_domain> such as "https://www.givengine.org" we set in the GIVE-Hub.
- `-h`: show usage help


### 8. `list_tracks.sh`
List existing data tracks in a reference genome. It will show tree structure of the relationships of the track groups and tracks.

If -a is set as "true", it will ignore other arguments and output all the reference genomes and associated track groups and tracks. If -a is not set as "true", the 3 arguments -r, -g, -t will determine the output structure level. If only set -r, it will output all the track group names and associated track names in the -r specified reference genome database. If set both -r and -g, then it wll list all the track names of -r -g specified track group. If set all the -r -g -t arguments, it will output the detail info of -r -g -t specified track.

```bash
bash list_tracks.sh [-u <mysqlu>] [-p <mysqlp>] [-a <true>] [-r <ref>] [-g <group_name>] [-t <track_name>]
```

- `-u <mysqlu>`: (Required) MySQL account name
- `-p <mysqlp>`: (Optional) MySQL account password. Without '-p', the prompt will ask you to input password.
- `-a <true>`: (Optional) If set as "true", it will list all the tracks in all the reference genomes with database tree structure.
- `-r <ref>`: (Optional) ref genome database name.
- `-g <group_name>`: (Optional) track group name.
- `-t <track_name>`: (Optional) track name.
- `-h`: show usage help

### 9. `remove_data.sh`
Remove specified data track, track group or the whole reference genome.

 ```bash
 bash remove_data.sh [-u <mysqlu>] [-p <mysqlp>] [-r <ref>] [-g <group_name>] [-t <track_name>] [-a <CONFIRM>]
```

- `-u <mysqlu>`: (Required) MySQL account name
- `-p <mysqlp>`: (Optional) MySQL account password. Without '-p', the prompt will ask you to input password.
- `-r <ref>`: (Required) ref genome database name. The ref genome database must has been initialized. If not, please use initial_ref.sh to do it.
- `-t <track_name>`: (Required for removing a track) The first priority of removing data. If -t argument was supplied, only the track will be removed. -g and -a arguments will be ignored.
- `-g <group_name>`: (Required for removing a track group) The second priority of removing data. If -t was not used and -g was supplied, then the track group will be removed. -a arguments will be ignored.
- `-a <TRUE>`: (Required for removing the whole ref genome database) The third priority of removing data. If -t and -g were not used, and "-a CONFIRM" was set, then the whole ref genome database (the name is provided with -r argument) will be removed, including all the tracks belong to it. It's dangerous, be careful.
- `-h`: show usage help




||||
| --- | --- | --- |
| [← 2.2. Details of GIVE Web Components](2.2-webComponents.md) | [↑ Index](Readme.md) | [3.2. MySQL commands for data management →](3.2-dataSource.md) |
