||||
| --- | --- | --- |
|[← 4.3.2. Data conversion from Hi-C contact matrices](4.3.2-HiC-conversion.md)| [↑ Index](Readme.md) | -- |

# Gene Annotation Track

Gene annotation track shows the locations and structures of genes and transcripts. There are many kinds of formats and sources of gene annotations, such as UCSC known gene table, GTF, GFF, bed, *etc*. Currently, GIVE only supports UCSC known gene table format. GTF/GFF support is coming in next update. Gene annotation track is set as `genePred` type in GIVE data source.

## Data Format

### UCSC known gene table format

UCSC known gene table format is used by UCSC known gene dataset. The UCSC Known Genes dataset is constructed by a fully automated process, based on protein data from Swiss-Prot/TrEMBL (UniProt) and the associated mRNA data from Genbank. It's a `Tab` separated 12 column text file format. Here, we describe the content of each column.

> __name__	Name of gene   
> __chrom__	Reference sequence chromosome or scaffold  
> __strand__	+ or - for strand  
> __txStart__	Transcription start position (or end position for minus strand item)  
> __txEnd__	Transcription end position (or start position for minus strand item)  
> __cdsStart__	Coding region start (or end position if for minus strand item)  
> __cdsEnd__	Coding region end (or start position if for minus strand item)  
> __exonCount__	Number of exons  
> __exonStarts__	Exon start positions (or end positions for minus strand item)  
> __exonEnds__	Exon end positions (or start positions for minus strand item)  
> __proteinID__	UniProt display ID, UniProt accession, or RefSeq protein ID  
> __alignID__	Unique identifier (GENCODE transcript ID for GENCODE Basic)  

