.getBlockStarts()<link rel="import" href="track-dom-behavior.html">
<link rel="import" href="../../../genemo-data-components/gene-object/gene-object.html">
<dom-module id="bed-track-dom">
  <template>
  </template>
  <script>
    var GIVe = (function (give) {
      'use strict'

      give.BedTrackDOM = Polymer({
        is: 'bed-track-dom',

        behaviors: [
          give.TrackDOMBehavior
        ],

        properties: {

        // data structure for bed tracks:
        // data has one key (chrom), all other keys will be deleted upon changing chromosome
        // data[chrom] is an array of Genes (for now)
        //    Future plan: arrays
        //    outside indices are start coordinates of every gene
        //    inside indices are used if multiple genes start at the same location
        //    NOTICE: start is always smaller than end, if the strand is negative, start is TTS, end is TSS
        //     values in the inner array are Gene objects

          honorItemRGB: {
          // whether this track honors itemRGB values
            type: Boolean,
            value: false
          }

        },

        created: function () {
          this.GENE_MARGIN = 10
          this.GENE_NOTEXT_MARGIN = 2
          this.ADAPTIVE_MAXLINES = 12    // limit to downgrade visibility
          this.TRIANGLE_FILL = 0xFFFFFF    // the fill color for the triangles (indicating clipped content)
          this.FORECOLOR_INDEX = 0      // the color index for fore color
        },

      // ****** customized methods below ******

        trackImpl: function (track, prop) {
          if (prop.hasOwnProperty('honorItemRGB')) {
            this.honorItemRGB = prop.honorItemRGB
          }
        },

        drawData: function () {
        // this is to draw everything from this.data to the svg
        // Steps:
        //     put genes into lines (pack display)
        //    draw genes out line by line

        // clear text Margin svg
          this.clear()

          this.changeVisibility(this.track.getSetting('visibility'), false)
          var numOfLines

          while (!(numOfLines = this.prepareLines()) &&
            this.activeVisibility > give.TrackObject.StatusEnum.VIS_NONE);

          if (this.track.data[this.mainSvg.viewWindow.chr]) {
            this.track.data[this.mainSvg.viewWindow.chr].traverse(this.mainSvg.viewWindow, function (gene) {
              if (this.activeVisibility > give.TrackObject.StatusEnum.VIS_COLLAPSED && gene.transcripts) {
                gene.transcripts.forEach(function (transcript) {
                  if (this.regionInWindow(transcript)) {
                    this.drawSingleGene(transcript,
                            this.activeVisibility > give.TrackObject.StatusEnum.VIS_DENSE ? null : 0,
                            (this.honorItemRGB && gene.itemRGB !== undefined) ? gene.itemRGB : null)
                  }
                }, this)
              } else {
                this.drawSingleGene(gene, this.activeVisibility > give.TrackObject.StatusEnum.VIS_DENSE ? null : 0,
                        (this.honorItemRGB && gene.itemRGB !== undefined) ? gene.itemRGB : null)
              }
            }, null, null, this, false)
          }

        // resize the heights
          this.updateLocationSize(null, null, null,
                    (numOfLines * (this.fullHeight + this.textLineGap) - this.textLineGap) * this.textSize)

        // add labels
          if (this.activeVisibility > give.TrackObject.StatusEnum.VIS_NONE) {
            this.drawShortLabel()
          }
        },

      // prepare genes

        prepareLines: function () {
        // calculate the x0 and x1 for every gene, (taking text width into consideration)
        // then update gene.y or transcript.y as needed
          var lineEnds = []    // arrays of line end coordinates
          if (this.track.data[this.mainSvg.viewWindow.chr]) {
            if (this.track.data[this.mainSvg.viewWindow.chr].traverse(this.mainSvg.viewWindow, function (gene) {
              var transcripts
              if (this.activeVisibility > give.TrackObject.StatusEnum.VIS_COLLAPSED && gene.transcripts) {
              // then transcripts needs to be drawn
                transcripts = gene.transcripts
              } else {
                transcripts = [gene]
              }
              return transcripts.every(function (transcript) {
              // calculate x0 and x1 for the gene
                if (!this.regionInWindow(transcript)) {
                  return true
                }
                var x0 = this.transformXCoordinate(transcript.getStartCoor(), true)
                var x1 = this.transformXCoordinate(transcript.getEndCoor(), true)
                var textAtLeft = false
                if (this.activeVisibility > give.TrackObject.StatusEnum.VIS_NOTEXT && transcript.getGeneName(true)) {
                  var newLabel = this.drawText(x0 - this.TEXT_MARGIN_GAP, this.Y_HIDDEN,
                               transcript.getGeneName(true), 'end')
                // move text to textMargin if out of bounds
                  if (this.textMargin && newLabel.getBBox().x < 0) {
                    x0 = 0
                    textAtLeft = true
                  } else {
                    x0 = newLabel.getBBox().x
                  }

                  this.removeElement(newLabel)
                }

                if (!lineEnds.some(function (lineEnd, index) {
                  if ((lineEnd.end <= x0 && !(lineEnd.textAtLeft && textAtLeft)) ||
                   this.activeVisibility <= give.TrackObject.StatusEnum.VIS_DENSE) {
                  // this gene can be fit into this line
                    transcript.y = index
                    lineEnd.end = x1 +
                      (this.activeVisibility > give.TrackObject.StatusEnum.VIS_NOTEXT
                        ? this.GENE_MARGIN : this.GENE_NOTEXT_MARGIN)
                    return true
                  }
                  return false
                }, this)) {
                // no empty lines, create a new line
                  if (this.track.getSetting('adaptive') && lineEnds.length >= this.ADAPTIVE_MAXLINES) {
                    this.changeVisibility(-1, true)
                    return false
                  }
                  transcript.y = lineEnds.length
                  lineEnds.push({
                    end: x1 +
                      (this.activeVisibility > give.TrackObject.StatusEnum.VIS_NOTEXT
                        ? this.GENE_MARGIN : this.GENE_NOTEXT_MARGIN),
                    textAtLeft: textAtLeft
                  })
                }
                return true
              }, this)
            }, null, null, this, true)) {
              return lineEnds.length > 0 ? lineEnds.length : 1
            } else {
              return false
            }
          }
          return 1
        },

        drawSingleGene: function (transcript, y, colorRGB, height, halfHeight, lineHeight) {
        // transcript: the transcript to be drawn
        // y: the y coordinate of the gene (line location)
        // if y is not a number then transcript.y will be used
        // height: the height of gene (notice that thin regions will have this.halfHeight * height)

          height = height || this.fullHeight * this.textSize * 0.8
          halfHeight = halfHeight || this.halfHeight
          lineHeight = lineHeight || (this.fullHeight + this.textLineGap) * this.textSize
          if (typeof (colorRGB) !== 'number') {
            colorRGB = this.colorSet[this.FORECOLOR_INDEX]
          }

          if (typeof (y) !== 'number' || isNaN(y)) {
            y = parseInt(transcript.y) * lineHeight || 0
          }

          var coor, blockStart, blockEnd
          var isThick = false
  //      // draw thin segments (if any)
  //      if(gene.thickStart > gene.getStart()) {
  //        coor = new give.ChromRegion(null, null, gene.chr, gene.getStart(), gene.thickStart, gene.strand);
  //        this.drawRectangle(coor, 0, y + height * 0.5 * halfHeight, height * halfHeight);
  //      }
  //      if(gene.thickEnd < gene.end) {
  //        coor = new give.ChromRegion(null, null, gene.chr, gene.thickEnd, gene.end, gene.strand);
  //      }

          if (transcript.getNumOfBlocks && transcript.getNumOfBlocks()) {
            blockEnd = transcript.getStart() + transcript.getBlockStarts()[0]
          // draw thick blocks and connecting lines
            for (var i = 0; i < transcript.getNumOfBlocks(); i++) {
              blockStart = transcript.getStart() + transcript.getBlockStarts()[i]
            // first draw connecting lines (intron, if any)
              if (blockEnd < blockStart) {
                coor = new give.ChromRegion({chr: transcript.chr,
                  start: blockEnd,
                  end: blockStart,
                  strand: transcript.strand})
                this.drawSpanningLine(coor, colorRGB, y, height)
              }
              blockEnd = blockStart + transcript.getBlockSizes()[i]
              if (transcript.thickStart < blockEnd && transcript.thickStart > blockStart) {
              // CDS start is in this block
                coor = new give.ChromRegion({chr: transcript.chr,
                  start: blockStart,
                  end: transcript.thickStart,
                  strand: transcript.strand})
                this.drawRectangle(coor, colorRGB, y + height * 0.5 * (1 - halfHeight), height * halfHeight, null, colorRGB)
                blockStart = transcript.thickStart
              }
              if (transcript.thickEnd < blockEnd && transcript.thickEnd > blockStart) {
              // CDS end is in this block
                coor = new give.ChromRegion({chr: transcript.chr,
                  start: blockStart,
                  end: transcript.thickEnd,
                  strand: transcript.strand})
                this.drawRectangle(coor, colorRGB, y, height, null, colorRGB)
                blockStart = transcript.thickEnd
              }

              isThick = (transcript.thickStart < blockEnd) && (transcript.thickEnd > blockStart)
              coor = new give.ChromRegion({chr: transcript.chr,
                start: blockStart,
                end: blockEnd,
                strand: transcript.strand})
              this.drawRectangle(coor, colorRGB, y + (isThick ? 0 : height * 0.5 * (1 - halfHeight)),
                      height * (isThick ? 1 : halfHeight), null, colorRGB)
            }
          } else {
          // no blocks
            blockStart = transcript.getStart()
            blockEnd = transcript.getEnd()
            if (typeof (transcript.thickStart) === 'number' || typeof (transcript.thickEnd) === 'number') {
              if (transcript.thickStart && transcript.thickStart > blockStart) {
              // CDS start is in this block
                coor = new give.ChromRegion({chr: transcript.chr,
                  start: blockStart,
                  end: transcript.thickStart,
                  strand: transcript.strand})
                this.drawRectangle(coor, colorRGB, y + height * 0.5 * (1 - halfHeight), height * halfHeight, null, colorRGB)
                blockStart = transcript.thickStart
              }
              if (transcript.thickEnd && transcript.thickEnd > blockStart) {
              // CDS end is in this block
                coor = new give.ChromRegion({chr: transcript.chr,
                  start: blockStart,
                  end: transcript.thickEnd,
                  strand: transcript.strand})
                this.drawRectangle(coor, colorRGB, y, height, null, colorRGB)
                blockStart = transcript.thickEnd
              }

              isThick = (transcript.thickStart < blockEnd) && (transcript.thickEnd > blockStart)
            } else {
              isThick = true
            }
            coor = new give.ChromRegion({chr: transcript.chr,
              start: blockStart,
              end: blockEnd,
              strand: transcript.strand})
            this.drawRectangle(coor, colorRGB, y + (isThick ? 0 : height * 0.5 * (1 - halfHeight)),
                    height * (isThick ? 1 : halfHeight), null, colorRGB)
          }

        // draw text
          if (this.activeVisibility > give.TrackObject.StatusEnum.VIS_NOTEXT) {
            var newLabel = this.drawText(this.transformXCoordinate(transcript.getStartCoor(), true) - this.TEXT_MARGIN_GAP,
                         y + 0.6 * height, transcript.getGeneName(true), 'end',
                         {style: 'fill: ' + this.rgbToHex(colorRGB)})

          // move text to textMargin if out of bounds
            if (this.textMargin && newLabel.getBBox().x < 0) {
              newLabel.setAttributeNS(null, 'x', this.textMargin)
              this.addElement(newLabel, this.textSvg)
            }
          }

        // draw triangle(s) indicating that the gene is not completely within the view
          if (this.transformXCoordinate(transcript.getStartCoor(), false) < 0) {
          // left triangle
            this.createRawPolygon(['0,' + (y + height * 0.5),
              height * 0.5 + ',' + y,
              height * 0.5 + ',' + (y + height * 0.5),
              height + ',' + y,
              height + ',' + (y + height),
              height * 0.5 + ',' + (y + height * 0.5),
              height * 0.5 + ',' + (y + height) ],
              {fill: this.TRIANGLE_FILL,
                stroke: colorRGB})
          }

          if (this.transformXCoordinate(transcript.getEndCoor(), false) > this.windowWidth) {
          // right triangle
            this.createRawPolygon([this.windowWidth + ',' + (y + height * 0.5),
              (this.windowWidth - height * 0.5) + ',' + y,
              (this.windowWidth - height * 0.5) + ',' + (y + height * 0.5),
              (this.windowWidth - height) + ',' + y,
              (this.windowWidth - height) + ',' + (y + height),
              (this.windowWidth - height * 0.5) + ',' + (y + height * 0.5),
              (this.windowWidth - height * 0.5) + ',' + (y + height) ],
              {fill: this.TRIANGLE_FILL,
                stroke: colorRGB})
          }
        },

        drawSpanningLine: function (region, colorRGB, y, height) {
          height = height || 1
          y = y || 0

          var svgToDraw = this.mainSvg
          var windowToDraw = svgToDraw.viewWindow

          if (windowToDraw.overlaps(region) > 0) {
            var x0 = this.transformXCoordinate(region.getStartCoor(), true)
            var x1 = this.transformXCoordinate(region.getEndCoor(), true)
            this.drawLine(x0, y + 0.5 * height, x1, y + 0.5 * height, colorRGB)
            this.drawStrandArrows(x0, y, x1, y + height, region.getStrand(), colorRGB)
          }
        }

      })

      return give
    })(GIVe || {})
  </script>
</dom-module>
