# ChIA-PET long-range chromatin interactions

This genome browser demo presents 15 datasets of ChIA-PET long-range chromatin interactions along with human genome assembly GRCh37 (hg19) for comparative studies. ChIA-PET data were generated by [**ENCODE** Phase 2](https://www.encodeproject.org/matrix/?type=Experiment&assay_title=ChIA-PET&assembly=hg19&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&award.rfa=ENCODE2&files.file_type=bed+bed12). The ChIA-PET chromatin interaction data includes 8 experiments in 5 cell lines (K562, MCF-7, HeLa-S3, HCT116 and NB4) with 3 kinds of target genes (POLR2A, CTCF and ESR1).

![fig](./GIVE_demo2_chiapet.PNG)

## Data preparation for GIVE
We used the ENCODE ChIA-PET long-range chromatin interactions bed (bed12) format data, which can be download from [**ENCODE** Experiment Matrix](https://www.encodeproject.org/matrix/?type=Experiment&assay_title=ChIA-PET&assembly=hg19&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&award.rfa=ENCODE2&files.file_type=bed+bed12). ALternatively, you can use [`batch_download.txt`](./) with following shell command to download all the data.

```bash
## run the following command in linux shell 
xargs -n 1 curl -O -L < batch_download.txt
```
The following table shows all the 15 bed format data sets we can get from [**ENCODE** Phase 2](https://www.encodeproject.org/matrix/?type=Experiment&assay_title=ChIA-PET&assembly=hg19&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&award.rfa=ENCODE2&files.file_type=bed+bed12).


|DataFile|Experiment Accession|Target gene|Cell line|Description|
| --- |--- |--- |--- |--- |
|ENCFF001THT.bed.gz|ENCSR000BZX|POLR2A|HCT116|POLR2A ChIA-PET on human HCT-116|
|ENCFF001THU.bed.gz|ENCSR000BZW|POLR2A|HeLa-S3|POLR2A ChIA-PET on human HeLa-S3|
|ENCFF001THV.bed.gz|ENCSR000CAC|CTCF|K562|CTCF ChIA-PET on human K562|
|ENCFF001THW.bed.gz|ENCSR000BZY|POLR2A|K562|POLR2A ChIA-PET on human K562|
|ENCFF001THX.bed.gz|ENCSR000CAD|CTCF|MCF-7|CTCF ChIA-PET on human MCF-7|
|ENCFF001THY.bed.gz|ENCSR000CAD|CTCF|MCF-7|CTCF ChIA-PET on human MCF-7|
|ENCFF001THZ.bed.gz|ENCSR000BZZ|ESR1|MCF-7|ESR1 ChIA-PET on human MCF-7|
|ENCFF001TIA.bed.gz|ENCSR000BZZ|ESR1|MCF-7|ESR1 ChIA-PET on human MCF-7|
|ENCFF001TIB.bed.gz|ENCSR000BZZ|ESR1|MCF-7|ESR1 ChIA-PET on human MCF-7|
|ENCFF001TIC.bed.gz|ENCSR000BZY|POLR2A|K562|POLR2A ChIA-PET on human K562|
|ENCFF001TID.bed.gz|ENCSR000CAA|POLR2A|MCF-7|POLR2A ChIA-PET on human MCF-7|
|ENCFF001TIE.bed.gz|ENCSR000CAA|POLR2A|MCF-7|POLR2A ChIA-PET on human MCF-7|
|ENCFF001TIF.bed.gz|ENCSR000CAA|POLR2A|MCF-7|POLR2A ChIA-PET on human MCF-7|
|ENCFF001TIG.bed.gz|ENCSR000CAB|POLR2A|NB4|POLR2A ChIA-PET on human NB4|
|ENCFF001TIJ.bed.gz|ENCSR000CAA|POLR2A|MCF-7|POLR2A ChIA-PET on human MCF-7|

The following table shows a sample of the bed12 format used in **ENCODE** datasets, and the full description can be found [here](https://genome.ucsc.edu/FAQ/FAQformat.html#format1). 

|chrom|chromStart|chromEnd|name|score|strand|thickStart|thickEnd|itermRgb|blockCount|blockSizes|blockStarts|
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
|chr1|3507144|3538145|chr1:3507144..3509308-chr1:3534421..3538145,3|300|.|3507144|3538145|255,0,0|2|2164,3724|0,27277|
|chr1|3507584|3520603|chr1:3507584..3509631-chr1:3518303..3520603,3|300|.|3507584|3520603|255,0,0|2|2047,2300|0,10719|
|chr1|761369|763199|chr1:761369..763199-chr8:182999..184804,2|200|.|761369|763199|255,0,0|1|1830|0|
|chr8|182999|184804|chr1:761369..763199-chr8:182999..184804,2|200|.|182999|184804|255,0,0|1|1805|0|

The first two rows in the sample show two intra-chromosome interactions, and the last two rows duplicately show one inter-chromosome interactions. So you may have found that we only need the `name` column to get all the interaction information. So we wrote a simple script [chiapet2give.sh](/.chiapet2give.sh) to convert the datasets to the GIVE supported interaction bed format (the format definiation can be found in [GIVE Manual](https://github.com/Zhong-Lab-UCSD/Genomic-Interactive-Visualization-Engine/blob/master/manuals/3-dataSource.md#adding-interaction-tracks)). Run following command we can get converted GIVE interaction bed files, which are named with a prefix `give_x_`.










